Libraries/Azure.psm1

```diff
@@ -396,6 +396,134 @@ Function Create-AllResourceGroupDeployments($SetupTypeData, $TestCaseData, $Dist
+
+                # Remove unmanged data disks
+                foreach ($uri in $VM.StorageProfile.DataDisks.Vhd.Uri) {
+                    $DataDiskStorageAccountName = $uri.Split('/')[2].Split('.')[0]
+                    $DataDiskContainerName = $uri.Split('/')[-2]
+                    $DataDiskName = $uri.Split('/')[-1]
+                    $DataDiskStorageAccount = $StorageAccounts | Where-Object { $_.StorageAccountName -eq $DataDiskStorageAccountName }
+                    Write-LogInfo "DeleteResourceGroup: Removing Data Disk  : $DataDiskName"
+                    $DataDiskStorageAccount | Remove-AzStorageBlob -Container $DataDiskContainerName -Blob $DataDiskName -Verbose -Force
+                }
+            }
+        }
+
+        # Remove the resource group which will remove all the remaining resources.
+        $MaxRetryAttemts = 10
+        Write-LogInfo "[Background Job] : Removing Resource Group : $ResourceGroupName"
+        $RemoveJob = Remove-AzResourceGroup -ResourceGroupName $ResourceGroupName -Force -AsJob
+        $isDeleting = (Get-AzResourceGroup -ResourceGroupName $ResourceGroupName).ProvisioningState -eq "Deleting"
+        while (!$isDeleting -and $MaxRetryAttemts -gt 0) {
+            $RgStatus = (Get-AzResourceGroup -ResourceGroupName $ResourceGroupName).ProvisioningState
+            $MaxRetryAttemts -= 1
+            $isDeleting = $RgStatus -eq "Deleting"
+            Write-LogInfo "[Background Job] : Removing Resource Group : $ResourceGroupName : $RgStatus. Remaining attempts - $MaxRetryAttemts"
+            Start-Sleep -Seconds 5
+        }
+        Write-LogInfo "[Background Job] : Removing Resource Group : $ResourceGroupName : $RgStatus"
+        $RemoveJob | Remove-Job -Force
+    }
+
+    # Check if VMs have unmanaged disks
+    $VMs = Get-AzVm -ResourceGroupName $RGName
+    $UnmanagedDiskCount = $VMs.StorageProfile.OSDisk.Vhd.Count + $VM.StorageProfile.DataDisks.Vhd.Count
+    if ($UnmanagedDiskCount -gt 0) {
+        if ($Global:XMLSecrets) {
+            # Unmanaged VHD cleanup is only possible when XML Secrets file is given, since Azure context files authentication doesn't work for background jobs.
+            # Azure powershell issue : https://github.com/Azure/azure-powershell/issues/9448
+            # Once this issue is resolved, we can use context file authentication for background cleanup.
+
+            Write-LogInfo "Detected $UnmanagedDiskCount unmanaged VHDs in $RGName"
+            $null = Start-Job -Name "DeleteResourceGroup-$RGName" `
+                -ScriptBlock $DeleteScriptBlock `
+                -ArgumentList $RGName,$WorkingDirectory,$Global:XMLSecrets
+            $VMStatus = (Get-AzVM -ResourceGroupName $RGName ).ProvisioningState | Get-Unique
+            $isDeleting = $VMStatus -eq "Deleting"
+
+            # Give at least 30 seconds to start all the VM cleanup operations.
+            # Timeout is adjusted based on number of VMs.
+            $MaxAttempts = 10 + $VMs.Count
```

Comment:
how long does it usually take to start all the cleanup operations? We don't want it to increase the execution time too much

Reply:
It usually takes less than 15 seconds.